Make `SESSION_COOKIE_SAMESITE` a setting

johnswanson opened the pull request:
Considerations here:

- I didn't want to break the existing behavior, which allows the env var to be case insensitive. This required a bit of weirdness in the getter/setter - we get the *raw* value, then normalize it to a keyword, then make sure it's one of the valid options.

- I'm not sure what the best practice is regarding i18n - I just went off the existing code, but I'm not entirely sure whether I need to also add translations to the `locales/*.po` files, and how I get those translations if so (err... chatGPT?).

- permissions: I'm not exactly sure how we decide between the `:settings-manager` and `:admin` settings. I went with `:settings-manager` here.

- I'm not sure whether the frontend has its own definitions of possible values for set-based settings, or if I need to expose that somehow?

The existing tests were good, easily adapted to the new behavior, and caught the above-mentioned backwards incompatibility issue with case insensitivity.

- Related: https://github.com/metabase/metabase-private/issues/182

johnswanson committed:
Make `SESSION_COOKIE_SAMESITE` a setting

Considerations here:

- I didn't want to break the existing behavior, which allows the env var to be case insensitive. This required a bit
of weirdness in the getter/setter - we get the *raw* value, then normalize it to a keyword, then make sure it's one of
the valid options.

- I'm not sure what the best practice is regarding i18n - I just went off the existing code, but I'm not entirely sure
whether I need to also add translations to the `locales/*.po` files, and how I get those translations if
so (err... chatGPT?).

- permissions: I'm not exactly sure how we decide between the `:settings-manager` and `:admin` settings. I went with
`:settings-manager` here.

- I'm not sure whether the frontend has its own definitions of possible values for set-based settings, or if I need to
expose that somehow?

The existing tests were good, easily adapted to the new behavior, and caught the above-mentioned backwards
incompatibility issue with case insensitivity.

noahmoss commented on src/metabase/server/middleware/session.clj:
```diff
@@ -68,6 +69,38 @@
+
+(defn- valid-session-cookie-samesite!
+  [value & [added-ex-data]]
+  (let [normalized-value (some-> value name str/lower-case keyword)]
```
You're getting a clj-kondo error here due to the use of `str/lower-case`—we discourage it in favor of `metabase.util/lower-case-en` which will have consistent results regardless of your site locale. (`str/lower-case` will have inconsistent results if your locale is Turkish which has bitten us before.)

Seems like it this linter rule was ignored in the previous code but I don't see a good reason for that over just using our custom function.

johnswanson replied:
Nice, thanks for the heads up - sorry for missing the linter error!

noahmoss commented on src/metabase/server/middleware/session.clj:
```diff
@@ -68,6 +69,38 @@
+  [value & [added-ex-data]]
+  (let [normalized-value (some-> value name str/lower-case keyword)]
+    (when-not (contains? possible-session-cookie-samesite-values normalized-value)
+      (throw (ex-info "Invalid value for session cookie samesite"
```
On the question of translation: an error string should be wrapped with the `metabase.util.i18n.tru` macro if there's any chance of it being user-facing on the frontend. `tru` will translate the string into the current user's locale, while `trs` will translate it into the site-wide locale.

In this case though, I think it would only show up in the backend logs, so it's fine to leave it untranslated. (Most backend logs are actually translated into the site locale currently, but I think the plan is to move away from that—see [this Slack discussion](https://metaboat.slack.com/archives/CKZEMT1MJ/p1697553976266779).)

noahmoss replied:
And no need to add translations yourself; we use POEditor to source translations from volunteers prior to each major release.

johnswanson replied:
I'm not positive, but I would assume that `(throw (ex-info "foo" {:http-status 400}))` is going to bubble up to the frontend - so maybe I should actually wrap this error message in `tru`?

Or actually, maybe this means it's not a good idea to throw one error that might bubble up to either logs or an HTTP error - because we need to translate one but not the other. So I'll actually pull those apart.

noahmoss reviewed (approved):
Awesome, looks great. Just a couple lint errors to resolve.

> I'm not exactly sure how we decide between the :settings-manager and :admin settings. I went with :settings-manager here.

"Settings manager" refers to users who aren't full admins but are granted access to (most of) the "Settings" page in the admin panel ([docs](https://www.metabase.com/docs/latest/permissions/application#settings-access)). Any settings exposed on these pages should have `:settings-manager` visibility, so I think it's the right choice here.

> I'm not sure whether the frontend has its own definitions of possible values for set-based settings

The settings framework doesn't have a formal concept of "set-based settings," just keyword settings with extra validation added on the backend. We don't have that many keyword settings overall tbh, but I _think_ the frontend does just hardcode values where it needs them. It would make sense though to add a `:values` option to `defsetting` and have it do validation automatically, and expose the values to the frontend when the setting is fetched to reduce duplication. Could you open a tech debt issue on this? Maybe it would be a good one for you to work on if you'd be interested.

johnswanson committed:
Throw localized exceptions when setting `session-cookie-samesite`

Since these exceptions might end up presented to the user, we want to translate them. The exceptions when we *read*
the setting, however, should not be translated (since they'll appear in the server logs).

noahmoss commented on src/metabase/config.clj:
```diff
@@ -56,8 +56,7 @@
 (def ^:private ee-app-defaults
-  {:embed-max-session-age      "1440"   ; how long a FULL APP EMBED session is valid for. One day, by default
-   :mb-session-cookie-samesite "lax"})
+  {:embed-max-session-age      "1440"}) ; how long a FULL APP EMBED session is valid for. One day, by default
```
nit - don't need the extra whitespace after removing the other map value

johnswanson replied:
🤦 thank you!

johnswanson committed:
Doh - whitespace

paoliniluis commented:
fixes https://github.com/metabase/metabase/issues/35364
